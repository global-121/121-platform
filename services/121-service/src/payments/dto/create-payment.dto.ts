import { ApiProperty } from '@nestjs/swagger';
import { IsNumber, IsOptional, IsString } from 'class-validator';

export class CreatePaymentDto {
  @ApiProperty({ example: 10 })
  @IsNumber()
  @IsOptional()
  public readonly amount?: number;

  @ApiProperty({
    example: 'Extra payment to registrations not included in first payment',
  })
  @IsString()
  @IsOptional()
  public readonly note?: string;
}
