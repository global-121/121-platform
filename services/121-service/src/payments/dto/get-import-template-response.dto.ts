export class GetImportTemplateResponseDto {
  public name: string;
  public template: string[];
}
