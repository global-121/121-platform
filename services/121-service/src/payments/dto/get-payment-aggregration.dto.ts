import { IsNumberString } from 'class-validator';

export class GetPaymentAggregationDto {
  @IsNumberString()
  programId: number;

  @IsNumberString()
  payment: number;
}
