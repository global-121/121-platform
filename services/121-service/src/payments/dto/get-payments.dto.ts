export interface GetPaymentsDto {
  payment: number;
  paymentDate: Date | string;
}
