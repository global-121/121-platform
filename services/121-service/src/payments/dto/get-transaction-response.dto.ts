import { ApiProperty } from '@nestjs/swagger';

import { TransactionStatusEnum } from '@121-service/src/payments/transactions/enums/transaction-status.enum';
import { RegistrationStatusEnum } from '@121-service/src/registration/enum/registration-status.enum';

export class GetTransactionResponseDto {
  @ApiProperty({ example: 1, type: 'number' })
  public id: number;
  @ApiProperty({
    example: '2023-09-28T08:00:10.363Z',
    type: 'string',
    format: 'date-time',
  })
  public created: Date;
  @ApiProperty({
    example: '2023-09-28T08:00:10.363Z',
    type: 'string',
    format: 'date-time',
  })
  public updated: Date;
  @ApiProperty({ example: 1, type: 'number' })
  public payment: number;
  @ApiProperty({ example: TransactionStatusEnum.success, type: 'string' })
  public status: TransactionStatusEnum;
  @ApiProperty({ example: 25, type: 'number' })
  public amount: number;
  @ApiProperty({ example: 'Something went wrong', type: 'string' })
  public errorMessage: string | null;
  @ApiProperty({ example: 'ironBank' })
  public programFinancialServiceProviderConfigurationName: string;
  @ApiProperty({ example: 111, type: 'number' })
  public registrationProgramId: number;
  @ApiProperty({ example: 111, type: 'number' })
  public registrationId: number;
  @ApiProperty({ example: RegistrationStatusEnum.included, type: 'string' })
  public registrationStatus: RegistrationStatusEnum;
  @ApiProperty({ example: '12345', type: 'string' })
  public referenceId: string;
  @ApiProperty({ example: 'Juan Garcia' })
  public registrationName?: string;
}
