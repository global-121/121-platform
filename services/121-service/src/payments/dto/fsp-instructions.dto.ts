import { ExcelFspInstructions } from '@121-service/src/payments/fsp-integration/excel/dto/excel-fsp-instructions.dto';

export type CsvInstructions = ExcelFspInstructions[];

export class FspInstructions {
  public data?: CsvInstructions | string;
  public fileType?: ExportFileType;
}

export enum ExportFileType {
  csv = 'csv',
  excel = 'excel',
  xml = 'xml',
}
