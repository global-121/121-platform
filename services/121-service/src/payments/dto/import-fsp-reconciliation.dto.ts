import { ApiProperty } from '@nestjs/swagger';
import { IsArray, IsNumber, IsOptional } from 'class-validator';

import { PaTransactionResultDto } from '@121-service/src/payments/dto/payment-transaction-result.dto';
import { StatusEnum } from '@121-service/src/shared/enum/status.enum';

export class ImportFspReconciliationArrayDto {
  @ApiProperty({ example: +24300000000 })
  @IsOptional()
  public phoneNumber?: string;

  @ApiProperty({ example: StatusEnum.success })
  @IsOptional()
  public status?: string;

  @ApiProperty({ example: 50 })
  @IsOptional()
  public amount?: string;

  @IsOptional()
  public paTransactionResult?: PaTransactionResultDto;
}

export class ImportFspReconciliationDto {
  @ApiProperty({ example: [] })
  @IsArray()
  public validatedArray: ImportFspReconciliationArrayDto[];

  @ApiProperty({ example: 1 })
  @IsNumber()
  public recordsCount: number;
}
