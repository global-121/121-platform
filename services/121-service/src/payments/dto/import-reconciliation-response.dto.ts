import { ApiProperty } from '@nestjs/swagger';

import { ReconciliationFeedbackDto } from '@121-service/src/payments/dto/reconciliation-feedback.dto';
import { AggregateImportResultDto } from '@121-service/src/registration/dto/bulk-import.dto';

export class ImportReconciliationResponseDto {
  @ApiProperty({
    type: [ReconciliationFeedbackDto],
    description: 'The import result',
  })
  importResult: ReconciliationFeedbackDto[];

  @ApiProperty({
    type: AggregateImportResultDto,
    description: 'The aggregate import result',
  })
  aggregateImportResult: AggregateImportResultDto;
}
