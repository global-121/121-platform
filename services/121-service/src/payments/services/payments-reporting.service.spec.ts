import { TestBed } from '@automock/jest';
import { Repository } from 'typeorm';

import { PaymentEntity } from '@121-service/src/payments/entities/payment.entity';
import { PaymentsReportingHelperService } from '@121-service/src/payments/services/payments-reporting.helper.service';
import { PaymentsReportingService } from '@121-service/src/payments/services/payments-reporting.service';
import { TransactionStatusEnum } from '@121-service/src/payments/transactions/enums/transaction-status.enum';
import { TransactionScopedRepository } from '@121-service/src/payments/transactions/transaction.scoped.repository';
import { ProgramRegistrationAttributeRepository } from '@121-service/src/programs/repositories/program-registration-attribute.repository';
import { MappedPaginatedRegistrationDto } from '@121-service/src/registration/dto/mapped-paginated-registration.dto';
import { RegistrationStatusEnum } from '@121-service/src/registration/enum/registration-status.enum';
import { RegistrationViewsMapper } from '@121-service/src/registration/mappers/registration-views.mapper';
import { RegistrationsPaginationService } from '@121-service/src/registration/services/registrations-pagination.service';

function createMockTransaction(
  referenceId: string,
  amount: number,
  status: TransactionStatusEnum,
) {
  return {
    id: 1,
    created: new Date(),
    updated: new Date(),
    paymentId: 1,
    paymentDate: new Date(),
    status,
    amount,
    errorMessage: null,
    programFspConfigurationName: 'someFsp',
    registrationProgramId: 2,
    registrationId: 3,
    registrationStatus: RegistrationStatusEnum.included,
    registrationReferenceId: referenceId,
  };
}

const mockTransactions = [
  createMockTransaction('101', 100, TransactionStatusEnum.success),
  createMockTransaction('102', 200, TransactionStatusEnum.success),
];

describe('PaymentsReportingService - getTransactions', () => {
  let service: PaymentsReportingService;
  let transactionScopedRepository: TransactionScopedRepository;
  let registrationPaginationService: RegistrationsPaginationService;
  let programRegistrationAttributeRepository: ProgramRegistrationAttributeRepository;
  let paymentsHelperService: PaymentsReportingHelperService;
  let paymentRepository: Repository<PaymentEntity>;

  beforeEach(async () => {
    const { unit, unitRef } = TestBed.create(
      PaymentsReportingService,
    ).compile();

    transactionScopedRepository = unitRef.get(TransactionScopedRepository);

    service = unit;
    registrationPaginationService = unitRef.get(RegistrationsPaginationService);

    jest.spyOn(transactionScopedRepository, 'getTransactions');
    jest.spyOn(
      registrationPaginationService,
      'getRegistrationViewsChunkedByReferenceIds',
    );

    registrationPaginationService = unitRef.get(RegistrationsPaginationService);
    programRegistrationAttributeRepository = unitRef.get(
      ProgramRegistrationAttributeRepository,
    );
    paymentsHelperService = unitRef.get(PaymentsReportingHelperService);
    paymentRepository = unitRef.get('PaymentEntityRepository');

    jest.spyOn<any, any>(service, 'getTransactions');
    jest.spyOn(paymentsHelperService, 'getSelectForExport');
    jest.spyOn(programRegistrationAttributeRepository, 'getDropdownAttributes');
    jest.spyOn(
      RegistrationViewsMapper,
      'replaceDropdownValuesWithEnglishLabel',
    );
    jest.spyOn(paymentRepository, 'findOne').mockResolvedValue({});
  });

  describe('getTransactionsByPaymentId', () => {
    it('should return transactions with names', async () => {
      // Arrange
      const programId = 1;
      const paymentId = 2;

      const mockRegistrationViews = [
        { referenceId: '101', name: 'John Doe' },
        { referenceId: '102', name: 'Jane Smith' },
      ] as MappedPaginatedRegistrationDto[];

      jest
        .spyOn(transactionScopedRepository, 'getTransactions')
        .mockResolvedValue(mockTransactions);

      jest
        .spyOn(
          registrationPaginationService,
          'getRegistrationViewsChunkedByReferenceIds',
        )
        .mockResolvedValue(mockRegistrationViews);

      // Act
      const result = await service.getTransactionsByPaymentId({
        programId,
        paymentId,
      });

      // Assert
      expect(result).toEqual([
        { ...mockTransactions[0], registrationName: 'John Doe' },
        { ...mockTransactions[1], registrationName: 'Jane Smith' },
      ]);
    });

    it('should throw 404 if payment does not exist', async () => {
      // Arrange
      const programId = 1;
      const paymentId = 999;
      // Simulate paymentRepository.findOne returning undefined
      (paymentRepository.findOne as jest.Mock).mockResolvedValue(undefined);

      // Act & Assert
      await expect(
        service.getTransactionsByPaymentId({ programId, paymentId }),
      ).rejects.toMatchSnapshot();
    });

    it('should return empty array when no transactions found', async () => {
      // Arrange
      const programId = 1;
      const paymentId = 2;

      jest
        .spyOn(transactionScopedRepository, 'getTransactions')
        .mockResolvedValue([]);

      // Act
      const result = await service.getTransactionsByPaymentId({
        programId,
        paymentId,
      });

      // Assert
      expect(transactionScopedRepository.getTransactions).toHaveBeenCalledWith({
        programId,
        paymentId,
      });
      expect(result).toEqual([]);
    });
  });

  describe('exportTransactionsUsingDateFilter', () => {
    it('should call getTransactions with correct params and return formatted fileDto', async () => {
      // Arrange
      const programId = 1;
      const fromDateString = '2024-01-01T00:00:00.000Z';
      const toDateString = '2024-01-31T23:59:59.000Z';
      const select = ['name', 'foo'];
      const transactions = [{ registrationReferenceId: 'ref1', name: 'Alice' }];
      const dropdownAttributes = [
        { name: 'foo', options: [{ value: 'bar', label: 'Bar' }] },
      ];
      const replacedRows = [
        { registrationReferenceId: 'ref1', name: 'Alice', foo: 'Bar' },
      ];
      const fileName = 'transactions_1_2024-01-01T00-00-00_2024-01-31T23-59-59';

      (paymentsHelperService.getSelectForExport as jest.Mock).mockResolvedValue(
        select,
      );
      (service as any).getTransactions.mockResolvedValue(transactions);
      (
        programRegistrationAttributeRepository.getDropdownAttributes as jest.Mock
      ).mockResolvedValue(dropdownAttributes);
      (
        RegistrationViewsMapper.replaceDropdownValuesWithEnglishLabel as jest.Mock
      ).mockReturnValue(replacedRows);

      (
        paymentsHelperService.createTransactionsExportFilename as jest.Mock
      ).mockReturnValue(fileName);

      // Act
      const result = await service.exportTransactionsUsingDateFilter({
        programId,
        fromDateString,
        toDateString,
      });

      // Assert
      expect(paymentsHelperService.getSelectForExport).toHaveBeenCalledWith(
        programId,
      );
      expect((service as any).getTransactions).toHaveBeenCalledWith({
        programId,
        select,
        fromDate: new Date(fromDateString),
        toDate: new Date(toDateString),
      });
      expect(
        programRegistrationAttributeRepository.getDropdownAttributes,
      ).toHaveBeenCalledWith({
        programId,
        select,
      });
      expect(
        RegistrationViewsMapper.replaceDropdownValuesWithEnglishLabel,
      ).toHaveBeenCalledWith({
        rows: transactions,
        attributes: dropdownAttributes,
      });
      expect(result.data).toEqual(replacedRows);
      expect(result.fileName).toEqual(fileName);
    });
  });

  describe('getPaymentEvents', () => {
    it('should throw 404 if payment does not exist', async () => {
      // Arrange
      const programId = 1;
      const paymentId = 999;
      (paymentRepository.findOne as jest.Mock).mockResolvedValue(undefined);

      // Act & Assert
      await expect(
        service.getPaymentEvents({ programId, paymentId }),
      ).rejects.toMatchSnapshot();
    });
  });
});
