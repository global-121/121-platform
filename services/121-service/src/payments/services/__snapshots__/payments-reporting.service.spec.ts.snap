// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PaymentsReportingService - getTransactions getPaymentEvents should throw 404 if payment does not exist 1`] = `[HttpException: Payment with ID 999 not found in program 1]`;

exports[`PaymentsReportingService - getTransactions getTransactionsByPaymentId should throw 404 if payment does not exist 1`] = `[HttpException: Payment with ID 999 not found in program 1]`;
